FreeTubeApp/FreeTube
# Fix `proxyVideos` setting for legacy formats & hls urls

# Fix `proxyVideos` setting for legacy formats & hls urls




## Pull Request Type


## Description
Currently, when the `proxy videos through invidious` setting is enabled, only DASH manifests are effected, but legacy formats and hls urls don't proxy.  This PR addresses this by replacing the `host` of stream urls with the current invidious instance and adding the appropriate `host` URL param. This is accomplished in a map for the `activeSourceList` which ensures all legacy formats go through the proxy when `proxyVideos` is true. Additionally, the links that come back in hls manifests from invidious don't always respect the `local` flag, so I added code to `videojs.Vhs.xhr.beforeRequest` to ensure that all requests that videojs makes go through the current invidious instance.

## Screenshots 
_(legacy format)_
|before|after|
|--|--|
|![before-legacy](https://github.com/FreeTubeApp/FreeTube/assets/106682128/df74c45b-e817-4427-a097-07d916718f15)|![after-legacy](https://github.com/FreeTubeApp/FreeTube/assets/106682128/350f0f31-2739-4979-80e9-aaeab6041d79)|

_(hls url)_
|before|after|
|--|--|
|![before-hls](https://github.com/FreeTubeApp/FreeTube/assets/106682128/54ca5443-d309-499d-95d6-91c31973a36c)|![after-hls](https://github.com/FreeTubeApp/FreeTube/assets/106682128/6ddfa4fc-015d-451d-9e1e-5c637a08352b)|

## Testing 
_(this should probably be tested with both `invdious` and `local` API preference)_
1. Enable `Proxy Videos Through Invidious`
2. Navigate to a video (ex: https://youtu.be/SqB0lUcqFbA)
3. Select `Use legacy formats`
4. Open the network tab and ensure the request is being sent to the currently selected invidious instance instead of something like `rr--3.googlevideo.com`
5. Navigate to a live stream with a hls url (ex: https://youtu.be/jfKfPfyJRdk)
6. Open the network tab and ensure the request is being sent to the currently selected invidious instance

_(as an added note, if you test this in a web browser using `yarn dev:web`, you don't have to check the network tab because the video will just fail to play if it is not proxied)_

## Desktop

 - OS: Windows 10
 - OS Version: Pro Version 21H2 Installed on ‎4/‎3/‎2022 OS build 19044.1889 Experience Windows Feature Experience Pack 120.2212.4180.0
- **FreeTube version:** 5295934

## Additional context
Currently, there is a much hackier version of this fix in FTCordova, but I just never got around to making an upstream PR about this issue until now even though it technically effects electron builds.
